### PR TITLE
ERC777Mintable implementation

### LINKED ISSUE #2024: ERC777Mintable Contract
**🧐 Motivation**

ERC20Mintable is available for ERC20 tokens. An ERC777Mintable contract would help the adoption of ERC777.

**📝 Details**

We will need to implement `ERC777Mintable.sol` in a similar fashion to `ERC20Mintable.sol` by passing the constructor values to the `_mint` function properly.

Here is an example implementation that could be used:

```solidity
import "./ERC777.sol";
import "../../access/roles/MinterRole.sol";

/**
 * @dev Extension of {ERC777} that adds a set of accounts with the {MinterRole},
 * which have permission to mint (create) new tokens as they see fit.
 *
 * At construction, the deployer of the contract is the only minter.
 */
contract ERC777Mintable is ERC777, MinterRole {
    /**
     * @dev See {ERC777-_mint}.
     *
     * Requirements:
     *
     * - the caller must have the {MinterRole}.
     */
    function mint(
        address operator,
        address account,
        uint256 amount,
        bytes memory userData,
        bytes memory operatorData
    ) public onlyMinter returns (bool) {
        _mint(operator, account, amount, userData, operatorData);
        return true;
    }
}
```

This implementation works as expected, even with the standard `Crowsale` and `MintedCrowdsale` contracts, so long as the `_deliverTokens` function is overridden to call `_mint` properly, like so:

```solidity
/**
     * @dev Overrides delivery by minting tokens upon purchase.
     * @param beneficiary Token purchaser
     * @param tokenAmount Number of tokens to be minted
     */
    function _deliverTokens(address beneficiary, uint256 tokenAmount) internal {
        require(
            ERC777Mintable(address(token())).mint(beneficiary, beneficiary, tokenAmount, "", ""),
                "MintedCrowdsale: minting failed"
        );
    }
```

It may be worth a separate PR to make `MintedCrowdsale` work with either token standard, or to migrate functionality for crowdsales completely over to ERC777 (not sure of the OpenZeppelin way of handling such upgrades).

### PR DESCRIPTION
**Pull request by awantoch**
Fixes #2024 

As previously discussed in #2024, this is an ERC777Mintable implementation that leverages the same principals as ERC721Mintable and ERC20Mintable, by inheriting MinterRole and passing constructor parameters from a high-level `mint` function to the underlying `_mint`.

I have tested this with other OpenZeppelin contracts as well and it works just as expected, including with Crowdsales.

Setting this PR to WIP since I'm still trying to get the tests to pass and need a bit of assistance. I can't seem to get passed the initial stage. I tried modeling my tests based off of what I see for ERC20 and ERC777 but I'm missing something. I'm getting the following error:

```
Error: Contract error: contract binary not set. Can't deploy new instance.
This contract may be abstract, not implement an abstract parent's methods completely
or not invoke an inherited contract's constructor correctly
```

I have run every build script I've seen so far, so I'm not sure what else I can compile or where I might be able to point to the contract's binary/ABI. I appreciate any assistance! 😄

### CODE REVIEW

**Comment by nventuro** [comment]
Not only did you add the relevant code, you also included the new contract in the documentation! :heart_eyes: Thanks a lot @awantoch!

Your tests are failing because your contracts are abstract. The reason for this is that, unlike `ERC20` and `ERC721`, `ERC777` has a constructor, which you must call from your derived contracts. Something similar happens with `ERC721Full`, see [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/mocks/ERC721MintableBurnableImpl.sol) how you need to explicitly call constructors that receive arguments.

In this case, I'd probably add a constructor to `ERC777Mintable` that calls `ERC777`'s, and then call that one from the mock contract.

**Comment by awantoch** [comment]
Happy to help @nventuro! 

That makes perfect sense, I updated my implementation to pass in the constructor properly. It's modeled the same way as ERC20Mintable and ERC721Mintable now, so users can now just add the `ERC777Mintable` to their contract's inheritance and rock and roll without any additional changes (except for maybe removing any existing custom mint logic from the code).

For example:

```solidity
pragma solidity ^0.5.0;

import "@openzeppelin/contracts/token/ERC777/ERC777.sol";

contract TestCoin is ERC777 {

    constructor(uint256 initialSupply, address[] memory defaultOperators) ERC777("TestERC777Mintable", "TEST", defaultOperators)
    public {
        _mint(msg.sender, msg.sender, initialSupply, "", "");
    }
}
```

Would become:

```solidity
pragma solidity ^0.5.0;

import "@openzeppelin/contracts/token/ERC777/ERC777.sol";
import "@openzeppelin/contracts/token/ERC777/ERC777Mintable.sol";

contract TestCoin is ERC777, ERC777Mintable {

    constructor(uint256 initialSupply, address[] memory defaultOperators) ERC777("TestERC777Mintable", "TEST", defaultOperators)
    public {
        mint(msg.sender, msg.sender, initialSupply, "", "");
    }
}
```

That, plus making sure EIP1820 was deployed and that data and operatorData were passed properly got things working smoothly as expected. Hopefully that'll do it! I appreciate you looking over this.

**Comment by awantoch** [comment]
@nventuro This PR should be good to go. Let me know if you have any additional comments/requests. Cheers!

File: test/token/ERC777/behaviors/ERC777Mintable.behavior.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+          beforeEach(async function () {
+            ({ logs: this.logs } = await this.token.mint(other, other, amount, data, operatorData, { from }));
+          });
+
+          it('mints the requested amount', async function () {
+            expect(await this.token.balanceOf(other)).to.be.bignumber.equal(amount);
+          });
+
+          it('emits a mint and a transfer event', async function () {
+            expectEvent.inLogs(this.logs, 'Transfer', {
```
Comment (nventuro) [inline test/token/ERC777/behaviors/ERC777Mintable.behavior.js]: We should check here the [`Minted`](https://docs.openzeppelin.com/contracts/2.x/api/token/erc777#IERC777-Minted-address-address-uint256-bytes-bytes-) event is emitted with the correct `data`, `operatorData`, etc. 

I wouldn't go into more detail than that (such as checking the receive hooks are called).

File: contracts/token/ERC777/README.adoc [hunk 2]
```diff
@@ -18,3 +18,7 @@ Additionally there are interfaces used to develop contracts that react to token
 {{IERC777Sender}}
 
 {{IERC777Recipient}}
+
+== Extensions
+
+{{ERC777Mintable}}
```
Comment (nventuro) [inline contracts/token/ERC777/README.adoc]: Can you add a note about Mintable in the opening paragraphs? Something like 

'Finally, the `ERC777Mintable` extension can be used to specify a set of addresses that are allowed to mint new tokens.'

**Comment by awantoch** [comment]
Thanks @nventuro ! I'll be able to update this soon, been busy with other things. Thanks for all the feedback -- will implement and notify you when ready!

**Comment by nventuro** [comment]
Closing, see [this comment][(https://github.com/OpenZeppelin/openzeppelin-contracts/issues/2024#issuecomment-580885987) for details on why and how we're moving forward.

**Pull request closed without merging**